### PR TITLE
Add Memory analysis for benchmarks

### PR DESCRIPTION
**Pull request by Nirothipan**
## Approach
- GC logs are collected and the logs are analyzed using GC viewer https://github.com/chewiebug/GCViewer
- From the Analysis, the GC Throughput is extracted and written to the initial CSV file which contains the benchmarks results ( like function name, the time taken, avg latency etc ).

## Automation tests
 N/A

## Security checks
 - Followed secure coding standards in http://wso2.com/technical-reports/wso2-secure-engineering-guidelines? yes
 - Ran FindSecurityBugs plugin and verified report? yes
 - Confirmed that this PR doesn't commit any keys, passwords, tokens, usernames, or other secrets? yes

### CODE REVIEW

File: benchmarks/benchmarktypes/benchmark-type-string.bal [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+    string str = <string>("This is string.");
+}
+
+public function benchmarkXmlValueOf() {
+    xml x = xml `<name>ram</name>`;
+    string str = io:sprintf("%s", x);
+}
+
+public function benchmarkStringJsonValueOf() {
+    json testJson = {"Hello":"World"};
```
Comment (anupama-pathirage) [inline benchmarks/benchmarktypes/benchmark-type-string.bal]: Formatting issue- should be { "Hello": "World" };

File: benchmarks/benchmark/benchmark-get-functions.bal [hunk 2]
```diff
@@ -84,3 +120,30 @@ function addFunctions() {
...
+    functions["benchmarkStringIntValueOf"] = benchmarktypes:benchmarkStringIntValueOf;
+    functions["benchmarkStringFloatValueOf"] = benchmarktypes:benchmarkStringFloatValueOf;
+    functions["benchmarkStringBooleanValueOf"] = benchmarktypes:benchmarkStringBooleanValueOf;
+    functions["benchmarkStringValueOf"] = benchmarktypes:benchmarkStringValueOf;
+    functions["benchmarkXmlValueOf"] = benchmarktypes:benchmarkXmlValueOf;
+    functions["benchmarkStringJsonValueOf"] = benchmarktypes:benchmarkStringJsonValueOf;
+    functions["benchmarkStringLength"] = benchmarktypes:benchmarkStringLength;
+    functions["benchmarkStringSplit"] = benchmarktypes:benchmarkStringSplit;
+    functions["benchmarkStringUnescape"] = benchmarktypes:benchmarkStringUnescape;
+
```
Comment (anupama-pathirage) [inline benchmarks/benchmark/benchmark-get-functions.bal]: Any reason for adding new line before the end "}"?

File: benchmarks/benchmark/benchmark-get-functions.bal [hunk 3]
```diff
@@ -84,3 +120,30 @@ function addFunctions() {
...
+}
```
Comment (anupama-pathirage) [inline benchmarks/benchmark/benchmark-get-functions.bal]: Do we need to import ballerina/io for this file?

File: benchmarks/benchmark/benchmark-get-functions.bal [hunk 4]
```diff
@@ -11,6 +11,18 @@ function getFunction(string name) returns (function()) {
 
 function addFunctions() {
```
Comment (chrishantha) [inline benchmarks/benchmark/benchmark-get-functions.bal]: Remove unnecessary blank lines.

File: benchmarks/executor [hunk 5]
```diff
@@ -7,18 +7,43 @@ chmod +x ${BAL_HOME}/bin/ballerina
...
+          export _JAVA_OPTIONS="-Xloggc:"$5gc_${functionName}.log""
+          ballerina run target/benchmark.balx $2 $3 ${functionName} | tr -d "\n" >> ${resultsFileName}
+          unset _JAVA_OPTIONS
+
+          ## check whether there are any GC events.
+          if  ((3 < "$(wc -l < target/gc_${functionName}.log)")); then
+
+             ###### Analyze GC Logs ######
+             gcSummaryFile=target/gc_summarry_${functionName}.csv
+             java -jar target/gcviewer-$6.jar target/gc_${functionName}.log ${gcSummaryFile}
```
Comment (chrishantha) [inline benchmarks/executor]: Assign all script arguments to variables with meaningful names at the top.

File: benchmarks/executor [hunk 6]
```diff
@@ -7,18 +7,43 @@ chmod +x ${BAL_HOME}/bin/ballerina
...
- do
-   ballerina run target/benchmark.balx $2 $3 ${functionName} >> ${resultsFileName}
- done
+    do
+          export _JAVA_OPTIONS="-Xloggc:"$5gc_${functionName}.log""
+          ballerina run target/benchmark.balx $2 $3 ${functionName} | tr -d "\n" >> ${resultsFileName}
+          unset _JAVA_OPTIONS
+
+          ## check whether there are any GC events.
+          if  ((3 < "$(wc -l < target/gc_${functionName}.log)")); then
```
Comment (chrishantha) [inline benchmarks/executor]: Flip the condition to make it more readable.

File: benchmarks/executor [hunk 7]
```diff
@@ -7,18 +7,43 @@ chmod +x ${BAL_HOME}/bin/ballerina
...
+
+             ###### Analyze GC Logs ######
+             gcSummaryFile=target/gc_summarry_${functionName}.csv
+             java -jar target/gcviewer-$6.jar target/gc_${functionName}.log ${gcSummaryFile}
+
+              # GC throughput
+              GC_throughput=$(awk -F "\"*;\"*" '{print $2}' ${gcSummaryFile} | awk NR==37 )
```
Comment (chrishantha) [inline benchmarks/executor]: It's better to use grep and grep for the exact name.

**Comment by Nirothipan** [comment]
@chrishanth made the requested chanegs

**Pull request merged**